pkg/controllers/management/rkeworkerupgrader/upgrade.go

```diff
@@ -0,0 +1,386 @@
+
+	logrus.Infof("cluster [%s] worker-upgrade: updating node [%s] with plan [%v]", cluster.Name, node.Name, np.Version)
+
+	updated, err := uh.nodes.Update(nodeCopy)
+	if err != nil {
+		return nil, fmt.Errorf("error updating node [%s] with plan %v", node.Name, err)
+	}
+
+	return updated, err
+}
+
+func (uh *upgradeHandler) updateNodePlanVersion(node *v3.Node, cluster *v3.Cluster) (*v3.Node, error) {
+	nodeCopy := node.DeepCopy()
+	nodeCopy.Status.NodePlan.Version = cluster.Status.NodeVersion
+	logrus.Infof("cluster [%s] worker-upgrade: updating node [%s] with plan version [%v]", cluster.Name,
+		node.Name, nodeCopy.Status.NodePlan.Version)
+
+	updated, err := uh.nodes.Update(nodeCopy)
+	if err != nil {
+		return nil, err
+	}
+	return updated, err
+
+}
+
+func (uh *upgradeHandler) getNodePlan(node *v3.Node, cluster *v3.Cluster) (*v3.RKEConfigNodePlan, error) {
+	var (
+		nodePlan *v3.RKEConfigNodePlan
+		err      error
+	)
+	if isNonWorkerOnly(node.Status.NodeConfig.Role) {
+		nodePlan, err = uh.nonWorkerPlan(node, cluster)
+	} else {
+		nodePlan, err = uh.workerPlan(node, cluster)
+	}
+	return nodePlan, err
+}
+
+func (uh *upgradeHandler) upgradeCluster(cluster *v3.Cluster, nodeName string) error {
+	clusterName := cluster.Name
+
+	uh.clusterLock.Lock(clusterName)
+	defer uh.clusterLock.Unlock(clusterName)
+
+	if !v3.ClusterConditionUpgraded.IsUnknown(cluster) {
+		clusterCopy := cluster.DeepCopy()
+		v3.ClusterConditionUpgraded.Unknown(clusterCopy)
+		v3.ClusterConditionUpgraded.Message(clusterCopy, "updating worker nodes")
+		var err error
+		cluster, err = uh.clusters.Update(clusterCopy)
+		if err != nil {
+			return err
+		}
+		logrus.Infof("cluster [%s] worker-upgrade: updated cluster for upgrading", clusterName)
+	}
+
+	nodes, err := uh.nodeLister.List(clusterName, labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	toPrepareMap, toProcessMap, upgradedMap, notReadyMap, filtered, upgrading, done := uh.filterNodes(nodes, cluster.Status.NodeVersion)
+
+	maxAllowed, err := getNum(cluster.Spec.RancherKubernetesEngineConfig.UpgradeStrategy.MaxUnavailableWorker, filtered)
+	if err != nil {
+		return err
+	}
+
+	if len(notReadyMap) > maxAllowed {
+		return fmt.Errorf("cluster [%s] worker-upgrade: not enough nodes to upgrade: nodes %v notReady %v maxUnavailable %v", clusterName, filtered, keys(notReadyMap), maxAllowed)
+	}
+
+	if len(notReadyMap) > 0 {
+		// nodes are already unavailable, update plan to reconcile
+		for name, node := range notReadyMap {
+			if node.Status.NodePlan.Version == cluster.Status.NodeVersion {
+				continue
+			}
+
+			if err := uh.processNode(node, cluster, "reconciling"); err != nil {
+				return err
+			}
+
+			logrus.Infof("cluster [%s] worker-upgrade: updated unavailable node [%s] to attempt upgrade", clusterName, name)
+		}
+	}
+
+	unavailable := upgrading + len(notReadyMap)
+
+	logrus.Debugf("cluster [%s] worker-upgrade: workerNodeInfo: nodes %v maxAllowed %v unavailable %v upgrading %v notReady %v "+
```

Comment:
Can we move this up to directly after maxAllowed (line 236) so we directly see what we are dealing with before processing any logic?

Probably add a `len()` as well as with many nodes you don't want to be counting the list of nodes to know how many there are being used for calculation.